server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1041,11 +1041,10 @@ public void externalRefresh(String source, ActionListener<Engine.RefreshResult>
     }
 
     /**
-     * Synchronously refreshes the engine for new search operations to reflect the latest
+     * Refreshes the engine for new search operations to reflect the latest
      * changes unless another thread is already refreshing the engine concurrently.
      */
-    @Nullable
-    public abstract RefreshResult maybeRefresh(String source) throws EngineException;
+    public abstract void maybeRefresh(String source, ActionListener<RefreshResult> listener);
```

Comment:
This is essentially a periodic refresh call (only used there), which needs to be async.